bouter/embedded/__init__.py

```diff
@@ -6,90 +6,91 @@
+        return [f"theta_{i:02}" for i in range(self.n_tail_segments)]
+
+    @decorators.cache_results(cache_filename="behavior_log")
+    def reconstruct_missing_segments(self, continue_curvature=None):
 
-    @tail_points_matrix.setter
-    def tail_points_matrix(self, matrix):
-        """Rewrite tail points in the tail dataframe
-        """
         columns = [f"theta_{i:02}" for i in range(self.n_tail_segments)]
-        self.behavior_log.loc[:, columns] = matrix
+        segments = self.behavior_log.loc[:, self.tail_columns].values.copy()
+
+        if "missing_n" in self.behavior_log.columns:
+            revert_pts = self.behavior_log["missing_n"].values
+        else:
+            revert_pts = None
+
+        # Revert if possible if continue_curvature is None:
+        if continue_curvature is None:
+            if revert_pts is not None:
+                fixed_segments = utilities.revert_segment_filling(
+                    segments,
+                    continue_curvature=continue_curvature,
+                    pts=revert_pts,
+                )
+                self.behavior_log.loc[:, columns] = fixed_segments
+
+        # Otherwise, use the parameter to do the filling:
+        else:
+            fixed_segments, missing_n = utilities.fill_out_segments(
+                segments,
+                continue_curvature=continue_curvature,
+                revert_pts=revert_pts,
+            )
+            self.behavior_log.loc[:, columns] = fixed_segments
+            self.behavior_log["missing_n"] = missing_n
 
-    @decorators.cache_results
-    def get_vigor(self, vigor_duration_s=0.05):
+        return self.behavior_log
+
+    @decorators.cache_results(cache_filename="behavior_log")
+    def compute_vigor(self, vigor_duration_s=0.05):
         """ Get vigor, the proxy of embedded fish forward velocity,
         a standard deviation calculated on a rolling window of tail curvature.
 
         :param vigor_duration: standard deviation window length in seconds
         :return:
         """
-        # TODO split in two methods so that it is set in the log df when
-        # loaded from cache (alternatively never write there)
-        if "vigor" in self.behavior_log.columns:
-            return self.behavior_log["vigor"]
-
         vigor_win = int(vigor_duration_s / self.behavior_dt)
         self.behavior_log["vigor"] = (
             self.behavior_log["tail_sum"]
             .interpolate()
             .rolling(vigor_win, center=True)
             .std()
         )
-        return self.behavior_log["vigor"]
-
-    def reconstruct_missing_segments(self, continue_curvature=0):
-        """ If the tail tip is not tracked throught the whole experiment
-        reconstruct the tail sum from the segments that are
-
-        """
-        utilities.fill_out_segments(
-            self.tail_points_matrix, continue_curvature=continue_curvature
-        )
+        return self.behavior_log
```

Comment:
compute_vigor returns the whole behavior log, because of caching?

Reply:
yes